atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/ComparingTimestampCache.java

```diff
@@ -0,0 +1,84 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb;
+
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import javax.annotation.Nullable;
+
+import com.palantir.atlasdb.cache.DefaultTimestampCache;
+import com.palantir.atlasdb.cache.OffHeapTimestampCache;
+import com.palantir.atlasdb.cache.TimestampCache;
+import com.palantir.atlasdb.persistent.api.PersistentTimestampStore;
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.logsafe.Preconditions;
+
+public final class ComparingTimestampCache implements TimestampCache {
+    private final TimestampCache first;
+    private final TimestampCache second;
+
+    public static TimestampCache comparingOffHeapForTests(
+            MetricsManager metricRegistry,
+            PersistentTimestampStore persistentTimestampStore) {
+        TimestampCache first = new DefaultTimestampCache(
+                metricRegistry.getRegistry(),
+                () -> AtlasDbConstants.DEFAULT_TIMESTAMP_CACHE_SIZE);
+
+        TimestampCache second = OffHeapTimestampCache.create(
+                persistentTimestampStore,
+                metricRegistry.getTaggedRegistry(),
+                () -> AtlasDbConstants.DEFAULT_TIMESTAMP_CACHE_SIZE
+        );
+
+        return new ComparingTimestampCache(first, second);
+    }
+
+    private ComparingTimestampCache(TimestampCache first, TimestampCache second) {
+        this.first = first;
+        this.second = second;
+    }
+
+
+    @Override
+    public synchronized void clear() {
+        first.clear();
+        second.clear();
+    }
+
+    @Override
+    public synchronized void putAlreadyCommittedTransaction(Long startTimestamp, Long commitTimestamp) {
+        first.putAlreadyCommittedTransaction(startTimestamp, commitTimestamp);
+        second.putAlreadyCommittedTransaction(startTimestamp, commitTimestamp);
+    }
+
+    @Nullable
+    @Override
+    public synchronized Long getCommitTimestampIfPresent(Long startTimestamp) {
+        Long firstCommitTimestamp = first.getCommitTimestampIfPresent(startTimestamp);
+        Long secondCommitTimestamp = second.getCommitTimestampIfPresent(startTimestamp);
+        if (firstCommitTimestamp == null || secondCommitTimestamp == null) {
+            return Stream.of(firstCommitTimestamp, secondCommitTimestamp)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElse(null);
+        }
+        Preconditions.checkState(firstCommitTimestamp.equals(secondCommitTimestamp),
+                "There is a bug in cache implementation");
```

Comment:
I would include as Args the start timestamp and commit timestamps reported by each of the services in this case, for easier debugging.